File: 107-binary-tree-level-order-traversal-ii/binary-tree-level-order-traversal-ii.py
# -*- coding:utf-8 -*-


# Given a binary tree, return the bottom-up level order traversal of its nodes' values. (ie, from left to right, level by level from leaf to root).
#
#
# For example:
# Given binary tree [3,9,20,null,null,15,7],
#
#     3
#    / \
#   9  20
#     /  \
#    15   7
#
#
#
# return its bottom-up level order traversal as:
#
# [
#   [15,7],
#   [9,20],
#   [3]
# ]
#
#


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    rlst = []

    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        self.rlst=[]
        self.levelList(root, 0)
        mx = max([item['hight'] for item in self.rlst])
        rst = [list() for _ in range(mx+1)]
        for item in self.rlst:
            rst[mx - item['hight']].append(item['val'])

        return rst

    def levelList(self, root, hight):
        if root:
            self.rlst.append({'val': root.val, 'hight': hight})
            hight = hight + 1
            if root.left:
                self.levelList(root.left, hight)
            if root.right:
                self.levelList(root.right, hight)

File: solutions/107-binary-tree-level-order-traversal-ii/binary-tree-level-order-traversal-ii.cpp
// Given a binary tree, return the bottom-up level order traversal of its nodes' values. (ie, from left to right, level by level from leaf to root).
//
//
// For example:
// Given binary tree [3,9,20,null,null,15,7],
//
//     3
//    / \
//   9  20
//     /  \
//    15   7
//
//
//
// return its bottom-up level order traversal as:
//
// [
//   [15,7],
//   [9,20],
//   [3]
// ]
//
//


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int>> ans{};
        unordered_map<int, vector<int>> m{};
        visit(root, &m, 0);
        reverse(m, &ans);
        return ans;
    }
    void visit(TreeNode* root, unordered_map<int, vector<int>>* m, int d) {
        if (root == nullptr) {
            return;
        }
        if (m->find(d) == m->end()) {
            m->emplace(d, vector<int>{root->val});
        }
        else {
            m->at(d).emplace_back(root->val);
        }
        visit(root->left, m, d + 1);
        visit(root->right, m, d + 1);
        
    }
    void reverse(unordered_map<int, vector<int>>& m, vector<vector<int>>* ans) {
        int s = m.size();
        for (int i = s - 1; i >= 0; i--) {
            ans->emplace_back(m[i]);
        }
    }
};
